File: src/main/java/wav/demon/StatCraft/Commands/SC/SCWorldChanges.java
package wav.demon.StatCraft.Commands.SC;

import com.mysema.query.Tuple;
import com.mysema.query.sql.SQLQuery;
import org.bukkit.command.CommandSender;
import wav.demon.StatCraft.Commands.ResponseBuilder;
import wav.demon.StatCraft.Querydsl.QPlayers;
import wav.demon.StatCraft.Querydsl.QWorldChange;
import wav.demon.StatCraft.StatCraft;

import java.util.List;

public class SCWorldChanges extends SCTemplate {

    public SCWorldChanges(StatCraft plugin) {
        super(plugin);
        this.plugin.getBaseCommand().registerCommand("worldchanges", this);
    }

    @Override
    public boolean hasPermission(CommandSender sender, String[] args) {
        return sender.hasPermission("statcraft.user.worldchanges");
    }

    @Override
    public String playerStatResponse(String name) {
        try {
            int id = plugin.getDatabaseManager().getPlayerId(name);
            if (id < 0)
                throw new Exception();

            SQLQuery query = plugin.getDatabaseManager().getNewQuery();
            QWorldChange w = QWorldChange.worldChange;
            Integer result = query.from(w).where(w.id.eq(id)).uniqueResult(w.amount.sum());

            return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("World Changes")
                    .addStat("Total", df.format(result == null ? 0 : result))
                    .toString();
        } catch (Exception e) {
            return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("World Changes")
                    .addStat("Total", String.valueOf(0))
                    .toString();
        }
    }

    @Override
    public String serverStatListResponse(int num) {
        SQLQuery query = plugin.getDatabaseManager().getNewQuery();
        QWorldChange w = QWorldChange.worldChange;
        QPlayers p = QPlayers.players;

        List<Tuple> list = query
                .from(w)
                .leftJoin(p)
                .on(w.id.eq(p.id))
                .groupBy(p.name)
                .orderBy(w.amount.sum().desc())
                .limit(num)
                .list(p.name, w.amount.sum());

        return topListResponse("World Changes", list);
    }
}

File: src/main/kotlin/com/demonwav/statcraft/commands/sc/SCWorldChanges.kt
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * https://www.demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.commands.sc

import com.demonwav.statcraft.StatCraft
import com.demonwav.statcraft.commands.ResponseBuilder
import com.demonwav.statcraft.querydsl.QPlayers
import com.demonwav.statcraft.querydsl.QWorldChange
import org.bukkit.command.CommandSender
import java.sql.Connection

class SCWorldChanges(plugin: StatCraft) : SCTemplate(plugin) {

    init {
        plugin.baseCommand.registerCommand("worldchange", this)
    }

    override fun hasPermission(sender: CommandSender, args: Array<out String>?) = sender.hasPermission("statcraft.user.worldchanges")

    override fun playerStatResponse(name: String, args: List<String>, connection: Connection): String {
        val id = getId(name) ?: return ResponseBuilder.build(plugin) {
            playerName { name }
            statName { "World Changes" }
            stats["Total"] = "0"
        }

        val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError

        val w = QWorldChange.worldChange
        val result = query.from(w).where(w.id.eq(id)).uniqueResult(w.amount.sum()) ?: 0

        return ResponseBuilder.build(plugin) {
            playerName { name }
            statName { "World Changes" }
            stats["Total"] = df.format(result)
        }
    }

    override fun serverStatListResponse(num: Long, args: List<String>, connection: Connection): String {
        val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError

        val w = QWorldChange.worldChange
        val p = QPlayers.players

        val list = query
            .from(w)
            .innerJoin(p)
            .on(w.id.eq(p.id))
            .groupBy(p.name)
            .orderBy(w.amount.sum().desc())
            .limit(num)
            .list(p.name, w.amount.sum())

        return topListResponse("World Changes", list)
    }
}
